libnavigation-ui/src/main/java/com/mapbox/navigation/ui/route/MapRouteLine.kt

```diff
@@ -402,14 +412,110 @@ internal class MapRouteLine(
         )
     }
 
-    fun updateDistanceRemainingCache(route: DirectionsRoute, distanceRemaining: Float) {
-        distanceRemainingCache.put(route, distanceRemaining)
+    private fun initPrimaryRoutePoints(route: DirectionsRoute) {
+        primaryRoutePoints = parseRoutePoints(route)
+        primaryRouteLineGranularDistances =
+            calculateRouteGranularDistances(primaryRoutePoints?.flatList ?: emptyList())
     }
 
-    fun inhibitVanishingPointUpdate(inhibitVanishingPointUpdate: Boolean) {
+    /**
+     * Decodes the route geometry into nested arrays of legs -> steps -> points.
+     *
+     * The first and last point of adjacent steps overlap and are duplicated.
+     */
+    private fun parseRoutePoints(route: DirectionsRoute): RoutePoints? {
+        val precision =
+            if (route.routeOptions()?.geometries() == DirectionsCriteria.GEOMETRY_POLYLINE) {
+                Constants.PRECISION_5
+            } else {
+                Constants.PRECISION_6
+            }
+
+        val nestedList = route.legs()?.map { routeLeg ->
+            routeLeg.steps()?.map { legStep ->
+                legStep.geometry()?.let { geometry ->
+                    PolylineUtils.decode(geometry, precision).toList()
+                } ?: return null
+            } ?: return null
+        } ?: return null
+
+        val flatList = nestedList.flatten().flatten()
+
+        return RoutePoints(nestedList, flatList)
+    }
+
+    /**
+     * Tries to find and cache the index of the upcoming [RouteLineDistancesIndex].
+     */
+    fun updateUpcomingRoutePointIndex(routeProgress: RouteProgress) {
+        ifNonNull(
+            routeProgress.currentLegProgress,
+            routeProgress.currentLegProgress?.currentStepProgress,
+            primaryRoutePoints
+        ) { currentLegProgress, currentStepProgress, completeRoutePoints ->
+            var allRemainingPoints = 0
+            /**
+             * Finds the count of remaining points in the current step.
+             *
+             * TurfMisc.lineSliceAlong places an additional point at index 0 to mark the precise
+             * cut-off point which we can safely ignore.
+             * We'll add the distance from the upcoming point to the current's puck position later.
+             */
+            allRemainingPoints += try {
+                TurfMisc.lineSliceAlong(
+                    LineString.fromLngLats(currentStepProgress.stepPoints ?: emptyList()),
+                    currentStepProgress.distanceTraveled.toDouble(),
+                    currentStepProgress.step?.distance() ?: 0.0,
+                    TurfConstants.UNIT_METERS
+                ).coordinates().drop(1).size
+            } catch (e: TurfException) {
+                0
+            }
```

Comment:
What about extracting this block of code into a `private` `fun`ction and give it a name based on the comment? This way `updateUpcomingRoutePointIndex` will be easier to read and understand and comment will become superfluous so it won't be necessary.

Reply:
All those small pieces of logic have no application outside of this main method's context, they will not be reused anywhere else, and splitting them up would only artificially increase the method count of the library. In my opinion, the comments already split the impl into readable chunks.

The same applies to all comments below.

Reply:
I'm gonna go ahead and merge, let's address later if needed.

Reply:
I don't agree. Normally [comments are a code smell](https://refactoring.guru/smells/comments), they get outdated quickly because code changes but comments don't 😬 so they start lying 🤯 

> The best comment is a good name for a method or class.

or a variable.

Our future selves would thank us!